File: Placement Sheet/Arrays/Initial Startups/Move_Zeros.java
import java.util.ArrayList;
import java.util.Scanner;

public class Move_Zeros {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter the size of the array: ");
        int size = input.nextInt();
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            System.out.printf("Enter %d number element of the array: ", i + 1);
            arr[i] = input.nextInt();
        }
        input.close();

        System.out.println("Before move zero: ");
        printArray(arr);

        int[] result = moveZerosBruteForce(arr);
        System.out.println("After move zero using brute-force method: ");
        printArray(result);

        System.out.print("After move zero using optimal approach: ");
        moveZeros(arr);
        printArray(arr);

    }

    private static void printArray(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    // Brute-Force
    private static int[] moveZerosBruteForce(int[] arr) {
        ArrayList<Integer> newArray = new ArrayList<>();
        ArrayList<Integer> zeroArray = new ArrayList<>();

        for (int i = 0; i < arr.length; i++) {
            if (arr[i] != 0) {
                newArray.add(arr[i]);
            } else {
                zeroArray.add(0);
            }
        }

        for (int i = 0; i < zeroArray.size(); i++) {
            newArray.add(zeroArray.get(i));
        }

        int[] result = new int[arr.length];
        for (int i = 0; i < newArray.size(); i++) {
            result[i] = newArray.get(i);
        }

        return result;
    }

    // Optimal
    private static void moveZeros(int[] arr) {
        int index = 0;

        for (int i = 0; i < arr.length; i++) {
            if (arr[i] != 0) {
                arr[index] = arr[i];
                index++;
            }
        }

        while (index < arr.length) {
            arr[index] = 0;
            index++;
        }
    }
}

File: Placement Sheet/Arrays/Initial Startups/Python/Move_Zeros.py
# moveZerosBruteForce

def moveZerosBruteForce(arr):
    if arr is None or len(arr) == 0:
        return
    non_zero_arr = []
    zero_arr = []

    for i in arr:
        if i != 0:
            non_zero_arr.append(i)
        if i == 0:
            zero_arr.append(i)

    non_zero_arr.extend(zero_arr)

    return non_zero_arr

# Optimal...
def moveZeros(arr):
    if arr is None or len(arr) == 0:
        return

    index = 0
    for i in range(0, len(arr)):
        if arr[i] != 0:
            arr[index] = arr[i]
            index += 1

    while index < len(arr):
        arr[index] = 0
        index += 1

size = int(input("Enter the size of the array: "))
arr = []

for i in range(0, size):
    data = int(input("Enter the element: "))
    arr.append(data)

print("Original Array: ", arr)

# arr = moveZerosBruteForce(arr)
moveZeros(arr)

print("Modified Array: ", arr)
